### PR TITLE
added override for hparams in load_from_ckpt

### LINKED ISSUE #1474: Customizing hparams after loading checkpoint
### Before asking:   
1. search the issues.   
2. search the docs.    

#### What is your question?
I'm wondering what the best practice for loading a model with different hparams than what is stored in the checkpoint?

I realize I could just load the model and set them afterwards e.g.: 
```
model = model.load_from_checkpoint(args.checkpoint_file) # Load model

# Set hparams etc..
model.hparams.arg1 = 0.0
model.hparams.arg2 = 1.0
 
```
But the problem is that my  model __init__ function depends on the hparams arg1 and arg2 so they're set too late.

I could also do 
```
checkpoint = torch.load(args.checkpoint_file)
checkpoint['hparams']['arg1'] = 0.0
checkpoint['hparams']['arg2'] = 1.0
model = model._load_state_dict(checkpoint)
```

The problem here is that i'm using the protected function _load_state_dict. Is there another way of solving this that i've missed? Or could we consider making _load_state_dict public?

### PR DESCRIPTION
**Pull request by williamFalcon**
Fixes #1474 

Enable the following... (ran into a bunch of issues with transfer learning when moving across clusters)... ugh.

I think @tullie has suggested something like this?

```python
# override some of the params with new values
MyLightningModule.load_from_checkpoint(
    PATH,
    hparam_overrides={'num_layers': 128, 'pretrained_ckpt_path': NEW_PATH}
)
```

### CODE REVIEW

**Comment by justusschock** [comment]
This doesn't work, if hparams is a nested dict, right?

**Comment by williamFalcon** [comment]
good point. i guess we can support replacing the nested hparams by passing nested updates here? match key by key within the tree

**Comment by justusschock** [comment]
Should work and probably is the same effort as having to flatten the dict.
While we're on it: Can tensorboard log nested hparams? If not, we may want to flatten the dict anyway after the network is initialized?

**Comment by williamFalcon** [comment]
you mean in the setter for hparams in the pl module?

**Comment by justusschock** [comment]
Yes maybe and undo also save it flattened (then you could acess the entries with something like `hparams['a.b']` in the example above) and unflatten it right before model init when laoding from checkpoint...

**Comment by williamFalcon** [comment]
@Borda is this how we get doc tests?

**Comment by williamFalcon** [comment]
@justusschock but if we flatten for the user it might be unexpected no?

1. maybe a better behavior is that on hparams setter in the module it becomes a namespace so that users don't have to use ['notation'].

2. for loggers we flatten all of them with the dot notation?

**Comment by awaelchli** [comment]
why is the override necessary? This already works by passing a dict to kwargs, and then it will be merged / updated with the one in the checkpoint and passed to the LightningModule. 

Except for the flattening, I would expect that to work already

**Comment by Borda** [comment]
> @Borda is this how we get doc tests?

What do you mean?

**Comment by williamFalcon** [comment]
I added the >>> on the docs.
Do I need something special to run as doc tests?

File: pytorch_lightning/core/saving.py [hunk 1]
```diff
@@ -48,6 +48,37 @@ def on_hpc_load(self, checkpoint: Dict[str, Any]) -> None:
...
 
+def update_hparams(hparams, updates):
+    """
+    Overrides hparams with new values
+
+    >>> hparams = {'a', {'b': 2}, 'c': 1}
+    >>> updates = {'a': {'b': 4}, 'c': 7}
+    >>> update_hparams(hparams, updates)
+    >>> assert hparams['a']['b'] == 4, 'update hparams failed'
+    >>> assert hparams['c'] == 7, 'update hparams failed'
```
Comment (awaelchli) [inline pytorch_lightning/core/saving.py:59]:
```suggestion
    >>> hparams = {'a', {'b': 2}, 'c': 1}
    >>> updates = {'a': {'b': 4}, 'c': 7}
    >>> update_hparams(hparams, updates)
    >>> hparams['a']['b']
    4
    >>> hparams['c']
    7
```
This is they way doctests are expected to test things (test + document usage at the same time) @williamFalcon

**Comment by Borda** [comment]
> I added the >>> on the docs.
> Do I need something special to run as doc tests?

it depends in it is python file, you do not need anything extra
if it is the rest file then you do not use >>> but put it to testcode
```
.. testcode::

    a = 1
```
cc: @awaelchli

File: pytorch_lightning/core/saving.py (hunk 1 above)
Comment (Borda) [reply pytorch_lightning/core/saving.py:59]: this is the way as it serves also as meaningful example... :]

**Pull request merged**